README.md

```diff
@@ -1,71 +1,157 @@
 
-### Updating dependencies
+```sh
+kubectl get -w route my-app -o yaml
+```
+
+## Common Failures & Diagnostics
+
+### Incorrect traffic behavior
+
+If you see that the rollout percentage for a Revision you just created stays at 0 or jumps to 100 directly, there can be a few potential culprits:
 
-`./hack/update-deps.sh`
+1. You have not applied any rollout policy.
+2. You have not annotated your Service/Configuration with the correct policy name.
+3. You are not using Knative Service, and your Route and Configuration have namespace and/or name mismatches.
+
+### Debugging tips
+
+You can view the logs of the KCD controller and/or webhook by using the following command:
+
+```sh
+kubectl logs -n knative-serving <pod_name_for_controller/webhook>
+```
 
-## Source Code Headers
+## Under-the-Hood Technicalities
 
-Every file containing source code must include copyright and license
-information. This includes any JS/CSS files that you might be serving out to
-browsers. (This is to help well-intentioned people avoid accidental copying that
-doesn't comply with the license.)
+### Continuous Delivery Webhook
 
-Apache header:
+The continuous delivery webhook serves two purposes:
 
-    Copyright 2020 Google LLC
+1. To make sure that the KService reconciler does not overwrite changes to the Route object, the webhook implements a method that intercepts and overwrites the Route with the desired routing spec before the Route reaches the K8s API server.
+2. To make sure that customized policies strictly adhere to all the required criteria, the webhook implements a method that valids any Policy object it receives, and rejects any errors.
 
-    Licensed under the Apache License, Version 2.0 (the "License");
-    you may not use this file except in compliance with the License.
-    You may obtain a copy of the License at
+You will likely stumble upon the continuous delivery webhook. Please do not modify or delete it.
 
-        https://www.apache.org/licenses/LICENSE-2.0
+### The PolicyState resource
 
-    Unless required by applicable law or agreed to in writing, software
-    distributed under the License is distributed on an "AS IS" BASIS,
-    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-    See the License for the specific language governing permissions and
-    limitations under the License.
+If you come across the `policystate` custom resource, please do not modify or delete it. The purpose of PolicyState objects is for the KCD controller to communicate the desired routing state to the continuous delivery webhook. Because the KCD controller and the webhook are not run in the same binary, PolicyStates are used to relay information. It is purely an internal mechanism that helps the webhook write the correct Route spec; any tampering with PolicyStates might cause undefined behavior.
```

Comment:
nit: prefer something like "This resource is owned by the KCD controller" instead of "do not modify or delete"

Reply:
fixed